src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,96 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_SIZE = 10;
+    public static final double ADD_SIE = 1.5;
+    private T[] mainData;
+    private int size;
+
+    public ArrayList() {
+        mainData = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        arrayCapacityControl();
+        mainData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index: " + index);
+        }
+        arrayCapacityControl();
+        System.arraycopy(mainData, index, mainData,index + 1,size - index);
+        mainData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return mainData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        mainData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement = mainData[index];
+        System.arraycopy(mainData, index + 1, mainData, index, size - index - 1);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && mainData[i] == null)
+                    || (element != null && element.equals(mainData[i]))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index: " + index
+                    + ", size: " + size);
+        }
+    }
+
+    private void arrayCapacityControl() {
```

Comment:
method name should sound like an action
```suggestion
    private void controlCapacity() {
```